core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsBackupResponse.java

```diff
@@ -5,116 +5,83 @@
+   public JGroupsBackupResponse(Map<XSiteBackup, CompletableFuture<ValidResponse>> syncBackupCalls,
          TimeService timeService) {
-      this.backupCalls = Objects.requireNonNull(backupCalls);
-      this.timeService = timeService;
-      sendTimeNanos = timeService.time();
-   }
-
-   @Override
-   public void waitForBackupToFinish() throws Exception {
-      long deductFromTimeout = timeService.timeDuration(sendTimeNanos, MILLISECONDS);
-      errors = new HashMap<>(backupCalls.size());
-      long elapsedTime = 0;
-      boolean hasSyncBackups = false;
-      for (Map.Entry<XSiteBackup, CompletableFuture<ValidResponse>> entry : backupCalls.entrySet()) {
-         XSiteBackup xSiteBackup = entry.getKey();
-
-         if (!xSiteBackup.isSync()) {
-            continue;
-         }
-
-         hasSyncBackups = true;
-         long timeout = xSiteBackup.getTimeout();
-         String siteName = xSiteBackup.getSiteName();
-
-         if (timeout > 0) { //0 means wait forever
-            timeout -= deductFromTimeout;
-            timeout -= elapsedTime;
-            if (timeout <= 0 && !entry.getValue().isDone() ) {
-               log.tracef("Timeout period %d exhausted with site %s", xSiteBackup.getTimeout(), siteName);
-               errors.put(siteName, newTimeoutException(xSiteBackup.getTimeout(), xSiteBackup));
-               addCommunicationError(siteName);
-               continue;
-            }
-         }
-
-         long startNanos = timeService.time();
-         Response response = null;
-         try {
-            response = entry.getValue().get(timeout, MILLISECONDS);
-         } catch (java.util.concurrent.TimeoutException te) {
-            errors.put(siteName, newTimeoutException(xSiteBackup.getTimeout(), xSiteBackup));
-            addCommunicationError(siteName);
-         } catch (ExecutionException ue) {
-            Throwable cause = ue.getCause();
-            cause.addSuppressed(new TraceException());
-            log.tracef(cause, "Communication error with site %s", siteName);
-            errors.put(siteName, filterException(cause));
-            addCommunicationError(siteName);
-         } finally {
-            elapsedTime += timeService.timeDuration(startNanos, MILLISECONDS);
-         }
-
-         log.tracef("Received response from site %s: %s", siteName, response);
+      assert !syncBackupCalls.isEmpty();
+      this.requests = new ArrayList<>(syncBackupCalls.size());
+      for (Map.Entry<XSiteBackup, CompletableFuture<ValidResponse>> entry : syncBackupCalls.entrySet()) {
+         this.requests.add(new XSiteRequest(entry.getKey(), entry.getValue()));
       }
-      if (hasSyncBackups) {
-         timeElapsedConsumer.accept(timeService.timeDuration(sendTimeNanos, MILLISECONDS));
+
+      this.sendTimeNanos = timeService.time();
+      this.timeService = timeService;
+      CompletableFuture<?>[] futures = requests.stream()
+            .filter(XSiteRequest::isSync)
```

Comment:
The constructor parameter is called `syncBackupCalls`, I'd rather assert they are really sync in the constructor instead of filtering them here.

You can also use `CompletionStages.aggregateCompletionStage()` instead of `CompletableFuture.allOf()`.

Reply:
the name is wrong XD

Reply:
what is the advantage of `CompletionStages.aggregateCompletionStage()`? saving the array allocation?

Reply:
The array too, but the streams also seem too much when most of the time there's just one backup site.

Reply:
ok. I admin the stream may be too much xD
I'll change!